### ISSUE TITLE
More targeting improvement.

### ISSUE THREAD

**Issue by tossaponk** [OpenModularTurretsTeam/OpenModularTurrets#236]
Edit: For some reason, this code only works on non-player entity. It seems that player prevPos does not set properly on MP server. I'll try to get around that and post the update code.

Edit2: New workaround implemented by manually tracking target if it is player entity. Otherwise, use old prevPos method.

From my previous aiming improvement. #76 

If the target is moving very fast. Then the projectile will slightly miss the target due to the estimation is not accounting the time needed to travel to new target position.

Also there's one more problem about the target.motionY always include acceleration from gravity even when target is standing on solid ground causing turret to shooting below the intended position. For example, a chicken standing on the ground at >20 block distance will not hit by the projectile because the turret think it's falling and compensate for target.motionY by shooting into the ground. (It's supposed to shoot along the line of sight. In the other word, shoot it in the face!)

My new approach now calculate speed from last frame compared to current frame and adjust the projectile speed a little to hit the new position at the exact same time as if it's shooting at the original position. (No more estimation, yay!)

---

Add these variable to TurretHead class
`private double targetLastX = 0;
private double targetLastY = 0;
private double targetLastZ = 0;`

---

Change these line from
https://github.com/OpenModularTurretsTeam/OpenModularTurrets/blob/master/src/main/java/openmodularturrets/tileentity/turrets/TurretHead.java#L243-L246
To

```
        // is there a target, and Has it died in the previous tick?
        if (target == null || target.isDead || this.getWorldObj().getEntityByID(
                target.getEntityId()) == null || ((EntityLivingBase) target).getHealth() <= 0.0F) {
            target = getTarget();

            // Start tracking if target is acquired
            if( target != null ) {
                targetLastX = target.prevPosX;
                targetLastY = target.prevPosY;
                targetLastZ = target.prevPosZ;
            }
        }
```

---

Insert into this line
https://github.com/OpenModularTurretsTeam/OpenModularTurrets/blob/master/src/main/java/openmodularturrets/tileentity/turrets/TurretHead.java#L256
With

```
        // Update target tracking (Player entity not setting motion data when moving via movement keys)
        double targetSpeedX = 0;
        double targetSpeedY = 0;
        double targetSpeedZ = 0;
        if (target != null && target instanceof EntityPlayerMP )
        {
            targetSpeedX = target.posX - targetLastX;
            targetSpeedY = target.posY - targetLastY;
            targetSpeedZ = target.posZ - targetLastZ;
            targetLastX = target.posX;
            targetLastY = target.posY;
            targetLastZ = target.posZ;
        }
```

---

Replace lines below with the code provided.
https://github.com/OpenModularTurretsTeam/OpenModularTurrets/blob/master/src/main/java/openmodularturrets/tileentity/turrets/TurretHead.java#L331-L349

```
            // Calculate speed from displacement from last tick (Or use tracking data if target is player)
            double speedX = target instanceof EntityPlayerMP ? targetSpeedX : target.posX - target.prevPosX;
            double speedY = target instanceof EntityPlayerMP ? targetSpeedY : target.posY - target.prevPosY;
            double speedZ = target instanceof EntityPlayerMP ? targetSpeedZ : target.posZ - target.prevPosZ;
            double d0 = target.posX - projectile.posX;
            double d1 = target.posY + (double) target.getEyeHeight() - projectile.posY;
            double d2 = target.posZ - projectile.posZ;
            double dist = MathHelper.sqrt_double(d0 * d0 + d1 * d1 + d2 * d2);
            float f1 = (float) dist * (0.2F * (getDistanceToEntity(target) * 0.04F));
            double accuracy = this.getTurretAccuracy() * (1 - TurretHeadUtil.getAccuraccyUpgrades(
                    base)) * (1 + TurretHeadUtil.getScattershotUpgrades(base));

            // Adjust new firing coordinate according to target speed
            double time = dist / (projectile.gravity == 0.00F ? 3.0 : 1.6);
            double adjustedX = d0 + speedX * time;
            double adjustedY = d1 + speedY * time;
            double adjustedZ = d2 + speedZ * time;

            // Calculate projectile speed scaling factor to travel to adjusted destination on time
            double dist2 = MathHelper.sqrt_double(adjustedX * adjustedX + adjustedY * adjustedY + adjustedZ * adjustedZ);
            float speedFactor = (float)(dist2 / dist);

            if (projectile.gravity == 0.00F) {
                projectile.setThrowableHeading(adjustedX, adjustedY,
                                               adjustedZ, 3.0F * speedFactor, (float) accuracy);
            } else {
                projectile.setThrowableHeading(adjustedX, adjustedY,
                                               adjustedZ, 1.6F * speedFactor, (float) accuracy);
            }
```

**Comment by SephirothWS** [comment]
Why not consider **forking the repository**, making these changes in your fork, then [https://github.com/OpenModularTurretsTeam/OpenModularTurrets/compare](Open a Pull Request) to make it easier for this to be implemented?

**Comment by tossaponk** [comment]
It's because I'm fixing a lot of bug on mods I'm currently using on my server. So my working folder is a mess. That's why all my previous issues are like this, lets someone who is already working on the project implement it.

**Comment by Keridos** [comment]
Will be released both in 1.7.10 and 1.10.2 Thanks a lot  @tossaponk
